src/services/resource.js

```diff
@@ -0,0 +1,1297 @@
+      .length === 0);
+
+  resourceActions.new = newExpandedResources.created
+    ?.map((resource) => ({
+      ...resource,
+    }));
+
+  // Recreate the remove dataset combining the removed and reduced where the reduced no longer
+  // references any sourceFields.
+  resourceActions.removed = [
+    ...removedReducedResources.removed,
+    ...resourceActions.reduced
+      .filter((resource) => resource.sourceFields.length === 0)
+      .reduce((resources, resource) => {
+        const exists = resources
+          .find((r) => r.genericId === resource.genericId);
+
+        if (exists) {
+          exists.resourceIds = [...exists.resourceIds, resource.resourceId];
+          return resources;
+        }
+
+        return [
+          ...resources,
+          {
+            ...resource,
+            resourceIds: [resource.resourceId],
+            resourceId: undefined,
+          },
+        ];
+      }, []),
+  ].reduce((resources, resource) => {
+    const exists = resources
+      .find((r) => r.genericId === resource.genericId);
+
+    if (exists) {
+      exists.resourceIds = [...exists.resourceIds, resource.resourceIds];
+      return resources;
+    }
+
+    return [...resources, resource];
+  }, []);
+
+  // Remove the empty sourceField records from the reduced dataset.
+  resourceActions.reduced = resourceActions.reduced
+    .filter((resource) => resource.sourceFields.length !== 0)
+    .map((resource) => ({
+      ...resource,
+    }));
+
+  return {
+    create: resourceActions.new,
+    update: [
+      ...(resourceActions.expanded || []),
+      ...(resourceActions.delta || []),
+      ...(resourceActions.reduced || []),
+    ],
+    destroy: !ignoreDestroy
+      ? resourceActions.removed
+      : [],
+  };
+};
+
+const getResourcesForModel = async (
+  model,
+  resourceTableForeignKey,
+  genericId,
+  includeAutoDetected = false,
+) => (
+  includeAutoDetected
+    ? model.findAll({
+      where: {
+        [resourceTableForeignKey]: genericId,
+      },
+      include: [
+        {
+          model: Resource,
+          as: 'resource',
+        },
+      ],
+    })
+    : model.findAll({
+      where: {
+        [resourceTableForeignKey]: genericId,
+        sourceFields: { [Op.contains]: 'resource' },
+      },
+      include: [
+        {
+          model: Resource,
+          as: 'resource',
+        },
+      ],
+    }));
+
+// Generic method for running the processFunction for an entity based on the data found at
+// the passed id
+const genericProcessEntityForResourcesById = async (
+  tableModel,
+  resourceTableModel,
+  resourceTableAs,
+  processFunction,
+  id,
+  urls,
+  resourceIds,
+  ignoreDestroy = false,
+) => {
+  const entity = await tableModel.findOne({
+    where: { id },
+    include: [{
+      model: resourceTableModel,
+      as: resourceTableAs,
+      include: [{
+        model: Resource,
+        as: 'resource',
+      }],
+    }],
+  });
+
+  return entity
+    && typeof entity === 'object'
```

Comment:
strings are objects :(

Reply:
"While the findAll() method returns an array even when you have only one matching row, the findOne() method always returns an Object or null when there’s no matching row."